routers/api/actions/artifacts.go

```diff
@@ -0,0 +1,506 @@
+	artifactXActionsResultsMD5Header = "x-actions-results-md5"
+)
+
+// const artifactXActionsResultsCRC64Header = "x-actions-results-crc64"
+
+const artifactRouteBase = "/_apis/pipelines/workflows/{taskID}/artifacts"
+
+func ArtifactsRoutes(goctx gocontext.Context, prefix string) *web.Route {
+	m := web.NewRoute()
+	m.Use(withContexter(goctx))
+
+	r := artifactRoutes{
+		prefix: prefix,
+		fs:     storage.ActionsArtifacts,
+	}
+
+	// retrieve, list and confirm artifacts
+	m.Get(artifactRouteBase, r.listArtifacts)
+	m.Post(artifactRouteBase, r.getUploadArtifactURL)
+	m.Patch(artifactRouteBase, r.comfirmUploadArtifact)
+
+	// handle container artifacts
+	m.Put(artifactRouteBase+"/{artifactID}/upload", r.uploadArtifact)
+	m.Get(artifactRouteBase+"/{artifactID}/path", r.getDownloadArtifactURL)
+	m.Get(artifactRouteBase+"/{artifactID}/download", r.downloadArtifact)
+
+	return m
+}
+
+// withContexter initializes a package context for a request.
+func withContexter(ctx gocontext.Context) func(next http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(resp http.ResponseWriter, req *http.Request) {
+			ctx := context.Context{
+				Resp: context.NewResponse(resp),
+				Data: map[string]interface{}{},
+			}
+			defer ctx.Close()
+
+			// action task call server api with Bearer ACTIONS_RUNTIME_TOKEN
+			// we should verify the ACTIONS_RUNTIME_TOKEN
+			authHeader := req.Header.Get("Authorization")
+			if len(authHeader) == 0 || !strings.HasPrefix(authHeader, "Bearer ") {
+				ctx.Error(http.StatusUnauthorized, "Bad authorization header")
+				return
+			}
+			authToken := strings.TrimPrefix(authHeader, "Bearer ")
+			task, err := actions.GetRunningTaskByToken(req.Context(), authToken)
+			if err != nil {
+				log.Error("Error runner api getting task: %v", err)
+				ctx.Error(http.StatusInternalServerError, "Error runner api getting task")
+				return
+			}
+			ctx.Data["task"] = task
+
+			ctx.Req = context.WithContext(req, &ctx)
+
+			next.ServeHTTP(ctx.Resp, ctx.Req)
+		})
+	}
+}
+
+type artifactRoutes struct {
+	prefix string
+	fs     storage.ObjectStorage
+}
+
+func (ar artifactRoutes) buildArtifactURL(taskID, artifactID int64, suffix string) (string, error) {
+	uploadURL := strings.TrimSuffix(setting.AppURL, "/") + strings.TrimSuffix(ar.prefix, "/") +
+		strings.ReplaceAll(artifactRouteBase, "{taskID}", strconv.FormatInt(taskID, 10)) +
+		"/" + strconv.FormatInt(artifactID, 10) + "/" + suffix
+	u, err := url.Parse(uploadURL)
+	if err != nil {
+		return "", err
+	}
+	return u.String(), nil
+}
+
+// getUploadArtifactURL generates a URL for uploading an artifact
+func (ar artifactRoutes) getUploadArtifactURL(ctx *context.Context) {
+	task, ok := ctx.Data["task"].(*actions.ActionTask)
+	if !ok {
+		log.Error("Error getting task in context")
+		ctx.Error(http.StatusInternalServerError, "Error getting task in context")
+		return
+	}
+	taskID := ctx.ParamsInt64("taskID")
+	if task.JobID != taskID {
```

Comment:
It really doesn't look good to me. The variable names are unclear.

What are taskID / task.ID / task.JobID ?